File: src/main/java/com/vjiki/music/service/PlaylistServiceImpl.java
package com.vjiki.music.service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.springframework.stereotype.Service;

import com.vjiki.music.dto.PlaylistResponse;
import com.vjiki.music.dto.PlaylistSongResponse;
import com.vjiki.music.dto.PlaylistWithSongsResponse;
import com.vjiki.music.dto.TagResponse;
import com.vjiki.music.entity.Playlist;
import com.vjiki.music.entity.PlaylistSong;
import com.vjiki.music.mapper.PlaylistMapper;
import com.vjiki.music.mapper.PlaylistSongMapper;
import com.vjiki.music.repository.PlaylistRepository;
import com.vjiki.music.repository.PlaylistSongRepository;

@Service
public class PlaylistServiceImpl implements PlaylistService {

    private final PlaylistRepository playlistRepository;
    private final PlaylistSongRepository playlistSongRepository;
    private final TagLookupService tagLookupService;

    public PlaylistServiceImpl(PlaylistRepository playlistRepository,
                               PlaylistSongRepository playlistSongRepository,
                               TagLookupService tagLookupService) {
        this.playlistRepository = playlistRepository;
        this.playlistSongRepository = playlistSongRepository;
        this.tagLookupService = tagLookupService;
    }

    @Override
    public List<PlaylistResponse> getPlaylistsByUserId(UUID userId) {
        return playlistRepository.findByUserIdWithUser(userId).stream()
                .map(PlaylistMapper::toResponse)
                .toList();
    }

    @Override
    public PlaylistWithSongsResponse getPlaylistWithSongs(UUID playlistId) {
        Playlist playlist = playlistRepository.findById(playlistId)
                .orElseThrow(() -> new IllegalArgumentException("Playlist not found: " + playlistId));

        List<PlaylistSong> playlistSongs = playlistSongRepository.findByPlaylistIdWithSong(playlistId);

        List<UUID> songIds = playlistSongs.stream()
                .map(ps -> ps.getSong() == null ? null : ps.getSong().getId())
                .filter(java.util.Objects::nonNull)
                .toList();
        Map<UUID, List<TagResponse>> tagsBySongId = tagLookupService.getTagsByTrackIds(songIds);

        List<PlaylistSongResponse> songsWithTags = new ArrayList<>(playlistSongs.size());
        for (PlaylistSong playlistSong : playlistSongs) {
            PlaylistSongResponse response = PlaylistSongMapper.toResponse(playlistSong);
            response.setTags(tagsBySongId.getOrDefault(response.getSongId(), List.of()));
            songsWithTags.add(response);
        }

        PlaylistResponse playlistResponse = PlaylistMapper.toResponse(playlist);

        return PlaylistWithSongsResponse.builder()
                .id(playlistResponse.getId())
                .userId(playlistResponse.getUserId())
                .userName(playlistResponse.getUserName())
                .userNickname(playlistResponse.getUserNickname())
                .name(playlistResponse.getName())
                .description(playlistResponse.getDescription())
                .coverUrl(playlistResponse.getCoverUrl())
                .type(playlistResponse.getType())
                .isPublic(playlistResponse.getIsPublic())
                .createdAt(playlistResponse.getCreatedAt())
                .modifiedAt(playlistResponse.getModifiedAt())
                .songs(songsWithTags)
                .build();
    }
}

File: src/main/kotlin/com/vjiki/music/service/PlaylistServiceImpl.kt
package com.vjiki.music.service

import com.vjiki.music.dto.PlaylistResponse
import com.vjiki.music.dto.PlaylistSongResponse
import com.vjiki.music.dto.PlaylistWithSongsResponse
import com.vjiki.music.mapper.PlaylistMapper.toResponse
import com.vjiki.music.mapper.PlaylistSongMapper.toResponse as playlistSongToResponse
import com.vjiki.music.repository.PlaylistRepository
import com.vjiki.music.repository.PlaylistSongRepository
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class PlaylistServiceImpl(
    private val playlistRepository: PlaylistRepository,
    private val playlistSongRepository: PlaylistSongRepository,
    private val tagLookupService: TagLookupService
) : PlaylistService {

    override fun getPlaylistsByUserId(userId: UUID): List<PlaylistResponse> {
        return playlistRepository.findByUserIdWithUser(userId).map { it.toResponse() }
    }

    override fun getPlaylistWithSongs(playlistId: UUID): PlaylistWithSongsResponse {
        val playlist = playlistRepository.findById(playlistId)
            .orElseThrow { IllegalArgumentException("Playlist not found: $playlistId") }

        val playlistSongs = playlistSongRepository.findByPlaylistIdWithSong(playlistId)
        val songs = playlistSongs.map { it.playlistSongToResponse() }

        val songIds = playlistSongs.map { it.song.id }
        val tagsBySongId = tagLookupService.getTagsByTrackIds(songIds)
        val songsWithTags = songs.map { ps ->
            ps.copy(tags = tagsBySongId[ps.songId] ?: emptyList())
        }

        val playlistResponse = playlist.toResponse()

        return PlaylistWithSongsResponse(
            id = playlistResponse.id,
            userId = playlistResponse.userId,
            userName = playlistResponse.userName,
            userNickname = playlistResponse.userNickname,
            name = playlistResponse.name,
            description = playlistResponse.description,
            coverUrl = playlistResponse.coverUrl,
            type = playlistResponse.type,
            isPublic = playlistResponse.isPublic,
            createdAt = playlistResponse.createdAt,
            modifiedAt = playlistResponse.modifiedAt,
            songs = songsWithTags
        )
    }
}

